src/encode_append.rs

```diff
@@ -107,46 +97,41 @@ where
 	I::IntoIter: ExactSizeIterator,
 {
 	let iter = iter.into_iter();
-	let input_len = iter.len();
-
-	// No data present, just encode the given input data.
-	if self_encoded.is_empty() {
-		crate::codec::compact_encode_len_to(&mut self_encoded, iter.len())?;
-		iter.for_each(|e| e.encode_to(&mut self_encoded));
-		return Ok(self_encoded);
-	}
-
-	let (new_len, encoded_len, encoded_new_len) = extract_length_data(&self_encoded, input_len)?;
-
-	let replace_len = |dest: &mut Vec<u8>| {
-		Compact(new_len).using_encoded(|e| {
-			dest[..encoded_new_len].copy_from_slice(e);
-		})
-	};
-
-	let append_new_elems = |dest: &mut Vec<u8>| iter.for_each(|a| a.encode_to(dest));
+	let items_to_append = iter.len();
 
-	// If old and new encoded len is equal, we don't need to copy the
-	// already encoded data.
-	if encoded_len == encoded_new_len {
-		replace_len(&mut self_encoded);
-		append_new_elems(&mut self_encoded);
-
-		Ok(self_encoded)
+	if vec.is_empty() {
+		crate::codec::compact_encode_len_to(&mut vec, items_to_append)?;
 	} else {
-		let size = encoded_new_len + self_encoded.len() - encoded_len;
-
-		let mut res = Vec::with_capacity(size + input_len * mem::size_of::<Item>());
-		unsafe { res.set_len(size); }
-
-		// Insert the new encoded len, copy the already encoded data and
-		// add the new element.
-		replace_len(&mut res);
-		res[encoded_new_len..size].copy_from_slice(&self_encoded[encoded_len..]);
-		append_new_elems(&mut res);
-
-		Ok(res)
+		let old_item_count = u32::from(Compact::<u32>::decode(&mut &vec[..])?);
+		let new_item_count = old_item_count
+			.checked_add(items_to_append as u32)
+			.ok_or("cannot append new items into a SCALE-encoded vector: length overflow due to too many items")?;
+
+		let old_item_count_encoded_bytesize = Compact::<u32>::compact_len(&old_item_count);
+		let new_item_count_encoded_bytesize = Compact::<u32>::compact_len(&new_item_count);
+
+		if old_item_count_encoded_bytesize == new_item_count_encoded_bytesize {
+			// The size of the length as encoded by SCALE didn't change, so we can just
+			// keep the old buffer as-is. We just need to update the length prefix.
+			Compact(new_item_count).using_encoded(|length_encoded|
+				vec[..old_item_count_encoded_bytesize].copy_from_slice(length_encoded)
+			);
+		} else {
+			// We can't update the length as the new length prefix will take up more
+			// space when encoded, so we need to move our data to make space for it.
+			let mut new_vec = Vec::with_capacity(
+				new_item_count_encoded_bytesize + new_item_count as usize * mem::size_of::<Item>()
```

Comment:
I dont quite understand the code. Why is it using `mem::size_of` instead of some `Encode::size_hint` or something?  
Or is there some assumed relation between in-memory and encoded size?

Reply:
Ah, you're right! Good catch! It's quite late here so I didn't notice that and just copied what the code was originally doing here, but this *is* incorrect.

In fact, now this is harmless since it's only used for `with_capacity` (as I removed the `unsafe`, but previously with the `unsafe` it looks like we had a potential out-of-bounds write...  yikes.

Reply:
I've simplified this. Not sure if there's a better way of doing this. (We can't really get the exact encoded bytesize of the new items since we haven't iterated over them yet.)